### PR TITLE
RDM-10918: Migrate Revoke Access APIs to new Access Control

### PR DESCRIPTION
**Pull request by mattnayler**
### JIRA link (if applicable) ###

[RDM-10918](https://tools.hmcts.net/jira/browse/RDM-10918) _"Migrate Revoke Access APIs to new Access Control."_

### Change description ###

Migrate the following revoke case-role APIs across to to new Access Control (Role Assignment Service)

*  Revoke Access V1, see [F-046](https://github.com/hmcts/ccd-data-store-api/blob/master/src/aat/resources/features/F-046%20-%20Revoke%20Access%20V1/F-046.feature)
* Remove Case-Assigned Users and Roles, see [F-111](https://github.com/hmcts/ccd-data-store-api/blob/master/src/aat/resources/features/F-111%20-%20Remove%20Case-Assigned%20Users%20and%20Roles/F-111.feature)

> NB: This PR is currently based of PR #1555 as there are a number of shared changes between the _"Remove Case-Assigned Users and Roles"_ processing and the _"Add Case-Assigned Users and Roles"_ migrated in [RDM-10924](https://tools.hmcts.net/jira/browse/RDM-10924).

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by mattnayler** [comment]
**Note**: This ticket was developed against a snapshot of the PR hmcts/am-role-assignment-service#871 which contains a version of the advance delete api, see [AM-1474](https://tools.hmcts.net/jira/browse/AM-1474) for details.

File: src/test/java/uk/gov/hmcts/ccd/data/casedataaccesscontrol/DefaultRoleAssignmentRepositoryIT.java [hunk 1]
```diff
@@ -238,6 +243,128 @@ private void validateRoleAssignment(RoleAssignmentResource actualRole,
...
+
+            // THEN
+            verify(1, postRequestedFor(urlMatching(deleteUrl))
+                .withRequestBody(matchingJsonPath("$.queryRequests[0].attributes.caseId[0]", equalTo(caseId1)))
+                .withRequestBody(matchingJsonPath("$.queryRequests[0].actorId[0]", equalTo(userId1)))
+                .withRequestBody(matchingJsonPath("$.queryRequests[0].roleType[0]", equalTo(RoleType.CASE.name())))
+                .withRequestBody(matchingJsonPath("$.queryRequests[0].roleName[0]", equalTo(role1))));
+        }
+
+        @DisplayName("should make a call to the delete role assignment by query end point for single case roles")
```
Comment (mattnayler) [inline src/test/java/uk/gov/hmcts/ccd/data/casedataaccesscontrol/DefaultRoleAssignmentRepositoryIT.java:295]: reword to reflect multiple case roles.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperation.java [hunk 2]
```diff
@@ -186,19 +197,49 @@ public void removeCaseUserRoles(List<CaseAssignedUserRoleWithOrganisation> caseU
...
-                        requestedAssignment.getCaseRole())
-                );
-            }
-        );
+            findAndFilterOnExistingCauRoles(cauRolesByCaseDetails);
+
+        if (applicationParams.getEnableAttributeBasedAccessControl()) {
+            List<RoleAssignmentsDeleteRequest> deleteRequests = new ArrayList<>();
+
+            // group by case
```
Comment (mattnayler) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperation.java:208]: reword: this is a "for each case" action not a "group by case" action

**Review by mattnayler — COMMENTED** [review]
Self review comments

File: src/main/java/uk/gov/hmcts/ccd/data/casedataaccesscontrol/MultipleQueryRequestResource.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.extern.jackson.Jacksonized;
+
+import java.util.List;
+
+@Builder
+@Getter
+@Jacksonized
+public class MultipleQueryRequestResource {
+
+    List<RoleAssignmentQuery> queryRequests;
```
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/ccd/data/casedataaccesscontrol/MultipleQueryRequestResource.java:14]: Should this be `private` access ?

File: src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/RoleAssignmentsDeleteRequest.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+@Builder
+@Getter
+public class RoleAssignmentsDeleteRequest {
+
+    String caseId;
+
+    String userId;
+
+    List<String> roleNames;
```
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/RoleAssignmentsDeleteRequest.java:16]: Should these be `private` access ?

File: src/test/java/uk/gov/hmcts/ccd/data/casedataaccesscontrol/DefaultRoleAssignmentRepositoryIT.java [hunk 5]
```diff
@@ -238,6 +243,128 @@ private void validateRoleAssignment(RoleAssignmentResource actualRole,
...
+        private final String deleteUrl = "/am/role-assignments/query/delete";
+
+        private final String caseId1 = "11111";
+        private final String caseId2 = "22222";
+
+        private final String userId1 = "12345";
+        private final String userId2 = "23456";
+
+        private final String role1 = "[ROLE1]";
+        private final String role2 = "[ROLE2]";
```
Comment (sullivangareth) [inline src/test/java/uk/gov/hmcts/ccd/data/casedataaccesscontrol/DefaultRoleAssignmentRepositoryIT.java:260]: If these are constants, should they be static ?

File: src/test/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperationTest.java [hunk 6]
```diff
@@ -1300,53 +1349,140 @@ void setUp() {
...
             verify(caseUserRepository, times(1)).revokeAccess(CASE_ID, USER_ID, CASE_ROLE);
+
+            verifyNoInteractions(roleAssignmentService);
+        }
+
+        @Test
+        @DisplayName("RA set to true, should remove single case user role")
+        void shouldRemoveSingleCaseUserRoleForRA() {
+
+            // ARRANGE
```
Comment (sullivangareth) [inline src/test/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperationTest.java:1381]: this is `Given, when, then` elsewhere - but this really is a minor comment !

File: src/test/java/uk/gov/hmcts/ccd/data/casedataaccesscontrol/DefaultRoleAssignmentRepositoryIT.java (hunk 1 above)
Comment (mattnayler) [reply src/test/java/uk/gov/hmcts/ccd/data/casedataaccesscontrol/DefaultRoleAssignmentRepositoryIT.java]: Resolved in 1ee8ce11700eb2cfa7b7763bbd00bd9846172417.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperation.java (hunk 2 above)
Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperation.java]: Resolved in 1ee8ce11700eb2cfa7b7763bbd00bd9846172417.

File: src/main/java/uk/gov/hmcts/ccd/data/casedataaccesscontrol/MultipleQueryRequestResource.java (hunk 3 above)
Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/ccd/data/casedataaccesscontrol/MultipleQueryRequestResource.java]: Resolved in 1ee8ce11700eb2cfa7b7763bbd00bd9846172417.

The previous use of Lombok's `@Value` generated a linting warning "@Value already marks non-static, package-local fields private".  However as I have switched to use Lombok's `@Getter` I should have re-adjusted the definition.  They now read as both `private` and `final`.

File: src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/RoleAssignmentsDeleteRequest.java (hunk 4 above)
Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/RoleAssignmentsDeleteRequest.java]: Resolved in 1ee8ce11700eb2cfa7b7763bbd00bd9846172417.

The previous use of Lombok's `@Value` generated a linting warning "@Value already marks non-static, package-local fields private".  However as I have switched to use Lombok's `@Getter` I should have re-adjusted the definition.  They now read as both `private` and `final`.

File: src/test/java/uk/gov/hmcts/ccd/data/casedataaccesscontrol/DefaultRoleAssignmentRepositoryIT.java (hunk 5 above)
Comment (mattnayler) [reply src/test/java/uk/gov/hmcts/ccd/data/casedataaccesscontrol/DefaultRoleAssignmentRepositoryIT.java]: Resolved in 1ee8ce11700eb2cfa7b7763bbd00bd9846172417.

File: src/test/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperationTest.java (hunk 6 above)
Comment (mattnayler) [reply src/test/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperationTest.java:1381]: Resolved in 1ee8ce11700eb2cfa7b7763bbd00bd9846172417.

Sorry, I use them both interchangeably.  Looking at the history I introduced ARRANGE/ACT/ASSERT under RDM-8606 over a year ago and then recently added GIVEN/WHEN/THEN in RDM-10924, #1555.  As the later is in shorter supply (14 vs 50) I have re-worded the recent additions.

**Review by sullivangareth — APPROVED** [review]

**Review by smathangi — APPROVED** [review]
LGTM

**Pull request merged**